chains/tendermint/chain.go

```diff
@@ -253,23 +265,124 @@ func (c *Chain) sendMsgs(msgs []sdk.Msg) (res *sdk.TxResponse, err error) {
 	// Build the transaction builder
 	txb, err := tx.BuildUnsignedTx(txf, msgs...)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	// Attach the signature to the transaction
-	err = tx.Sign(txf, c.config.Key, txb)
+	err = tx.Sign(txf, c.config.Key, txb, false)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	// Generate the transaction bytes
 	txBytes, err := ctx.TxConfig.TxEncoder()(txb.GetTx())
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	// Broadcast those bytes
-	return ctx.BroadcastTx(txBytes)
+	res, err := ctx.BroadcastTx(txBytes)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// transaction was executed, log the success or failure using the tx response code
+	// NOTE: error is nil, logic should use the returned error to determine if the
+	// transaction was successfully executed.
+	if res.Code != 0 {
+		c.LogFailedTx(res, err, msgs)
+		return res, false, nil
+	}
+
+	c.LogSuccessTx(res, msgs)
+	return res, true, nil
+}
+
+func prepareFactory(clientCtx sdkCtx.Context, txf tx.Factory) (tx.Factory, error) {
+	from := clientCtx.GetFromAddress()
+
+	if err := txf.AccountRetriever().EnsureExists(clientCtx, from); err != nil {
+		return txf, err
+	}
+
+	initNum, initSeq := txf.AccountNumber(), txf.Sequence()
+	if initNum == 0 || initSeq == 0 {
+		num, seq, err := txf.AccountRetriever().GetAccountNumberSequence(clientCtx, from)
+		if err != nil {
+			return txf, err
+		}
+
+		if initNum == 0 {
+			txf = txf.WithAccountNumber(num)
+		}
+
+		if initSeq == 0 {
+			txf = txf.WithSequence(seq)
+		}
+	}
+
+	return txf, nil
+}
+
+// protoTxProvider is a type which can provide a proto transaction. It is a
+// workaround to get access to the wrapper TxBuilder's method GetProtoTx().
+type protoTxProvider interface {
+	GetProtoTx() *txtypes.Tx
+}
+
+// BuildSimTx creates an unsigned tx with an empty single signature and returns
+// the encoded transaction or an error if the unsigned transaction cannot be
+// built.
+func BuildSimTx(txf tx.Factory, msgs ...sdk.Msg) ([]byte, error) {
+	txb, err := tx.BuildUnsignedTx(txf, msgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an empty signature literal as the ante handler will populate with a
+	// sentinel pubkey.
+	sig := signing.SignatureV2{
+		PubKey: &secp256k1.PubKey{},
+		Data: &signing.SingleSignatureData{
+			SignMode: txf.SignMode(),
+		},
+		Sequence: txf.Sequence(),
+	}
+	if err := txb.SetSignatures(sig); err != nil {
+		return nil, err
+	}
+
+	protoProvider, ok := txb.(protoTxProvider)
+	if !ok {
+		return nil, fmt.Errorf("cannot simulate amino tx")
+	}
+	simReq := txtypes.SimulateRequest{Tx: protoProvider.GetProtoTx()}
+
+	return simReq.Marshal()
+}
+
+// CalculateGas simulates the execution of a transaction and returns the
+// simulation response obtained by the query and the adjusted gas amount.
+func CalculateGas(
```

Comment:
I understand that these codes have been temporarily copied and will be removed later, is that correct?

Reply:
I think It's probably correct.
This change is based on the following: https://github.com/cosmos/relayer/pull/501/files#diff-ba13e4822ca5775807a70be387519276a6103762c525afb11068bdff525d7fcbR299
For now, I'd like to use cosmos/relayer codebase as much as possible to reduce the maintenance cost of tendermint support.

Reply:
:+1: